csharp/src/Ice/Internal/ConnectionFactory.cs

```diff
@@ -1595,6 +1612,13 @@ private void setState(int state)
                 {
                     return;
                 }
+
+                if (_maxConnections > 0 && _connections.Count == _maxConnections)
```

Comment:
Say the application calls `adapter.activate()` on an OA in the holding state.

If at this point `_connections.Count == _maxConnections`, the state of the adapter remains Holding. Should we throw an exception to notify the application `adapter.activate()` was not successful?

Moreover:
 - holding = don't accept new incoming connections AND don't read from existing (incoming) connections.
 - max connection reached = don't accept new incoming connections but keep reading from existing incoming connections

So I think we should actually transition from Holding to Active here, and resume reading from the existing connections. Just don't accept new connections as long as `_connections.Count == _maxConnections`.

Reply:
I agree.